src/proxy/hdrs/XPACK.cc

```diff
@@ -339,14 +339,22 @@ XpackDynamicTable::insert_entry(const char *name, size_t name_len, const char *v
   }
 
   // Make enough space to insert a new entry
-  uint64_t required_size = static_cast<uint64_t>(name_len) + static_cast<uint64_t>(value_len) + ADDITIONAL_32_BYTES;
+  uint64_t const required_storage_size = static_cast<uint64_t>(name_len) + static_cast<uint64_t>(value_len);
+  uint64_t const required_size         = required_storage_size + ADDITIONAL_32_BYTES;
   if (required_size > this->_available) {
     if (!this->_make_space(required_size - this->_available)) {
       // We can't insert a new entry because some stream(s) refer an entry that need to be evicted or the header is too big to
       // store. This is fine with HPACK, but not with QPACK.
       return {UINT32_C(0), XpackLookupResult::MatchType::NONE};
     }
   }
+  // Protect against overrun: ensure that our storage has space for the new
+  // header.
+  if (this->_storage.will_overrun(required_storage_size)) {
```

Comment:
Looks redundant and I don't think it can be true. How does this happen?

Reply:
It may be should be redundant, but this is the overrun that this patch addresses. Before this patch, this overrun happens because `_make_space` didn't expand _storage before this patch. Per your observation, that expansion is being added here.

In any case, rather than risk an overrun, I'm trying to add a defensive check. I understand your comment though. With the patch, it should be redundant. Maybe I should make this a release assert instead.

Reply:
I think a *release* assert should be avoided. If `will_overrun` returns true, something is wrong and it's unrecoverable state, so the connection has to be closed anyway, however, we should not abandon all the other established connections. I'm fine with a regular (non-release) assert.

Reply:
I appreciate the concern about release asserts, but guarding overruns is one place I suggest we should use them. Overruns cause mysterious crashes anyway (if we're lucky - otherwise we may be sending sensitive data to a host, 
 leading to a security issue), but it can be very difficult to find the culprit because the crash happens after the fact with little or no evidence as to who corrupted the memory. A release_assert at least makes the crash happen deterministically at the point of where the overrun will happen.

Reply:
I still think we shouldn't use release assert here, and I'd print logs and return an error so H2 impl can send GOAWAY frame with COMPRESSION_ERROR. But let's revisit this on #10082.